File: rust/crates/specgate-fixtures/src/multi_step.rs
// Multi-step state machine — two operations in sequence.
use specgate_annotations::*;

#[spec_setup("make_counter")]
fn make_counter() -> Counter {
    Counter { count: 0 }
}

struct Counter {
    #[spec_event]
    count: i32,
}

impl Counter {
    #[spec_operation("increment")]
    fn increment(&mut self) {
        self.count += 1;
    }

    #[spec_operation("decrement")]
    fn decrement(&mut self) {
        self.count -= 1;
    }
}

File: test/csharp/SpecGateFixtures/Stateful/MultiStep/multi_step.cs
using SpecGate.Annotations;

namespace SpecGateFixtures.Stateful.MultiStep;

/// <summary>Multi-step state machine: a counter driven through <c>increment</c> then <c>decrement</c>.</summary>
public class Counter
{
    /// <summary>The current count; each assignment is captured as a state mutation.</summary>
    [SpecEvent("count")]
    public int Count { get; set; }

    /// <summary>Builds a counter starting at zero.</summary>
    /// <returns>A new <see cref="Counter"/> with <see cref="Count"/> 0.</returns>
    [SpecSetup("increment")]
    public static Counter Make() => new() { Count = 0 };

    /// <summary>Increments the counter by one.</summary>
    [SpecOperation("increment")]
    public void Increment() => Count += 1;

    /// <summary>Decrements the counter by one.</summary>
    [SpecOperation("decrement")]
    public void Decrement() => Count -= 1;
}
